app/uk/gov/hmrc/eoricommoncomponent/frontend/controllers/ContactAddressController.scala

```diff
@@ -0,0 +1,129 @@
+
+import play.api.mvc._
+import uk.gov.hmrc.eoricommoncomponent.frontend.controllers.auth.AuthAction
+import uk.gov.hmrc.eoricommoncomponent.frontend.controllers.routes._
+import uk.gov.hmrc.eoricommoncomponent.frontend.domain.subscription.ContactAddressSubscriptionFlowPageGetEori
+import uk.gov.hmrc.eoricommoncomponent.frontend.domain.{LoggedInUserWithEnrolments, YesNo}
+import uk.gov.hmrc.eoricommoncomponent.frontend.forms.MatchingForms._
+import uk.gov.hmrc.eoricommoncomponent.frontend.forms.models.AddressViewModel
+import uk.gov.hmrc.eoricommoncomponent.frontend.models.Service
+import uk.gov.hmrc.eoricommoncomponent.frontend.services.cache.SessionCache
+import uk.gov.hmrc.eoricommoncomponent.frontend.services.{SubscriptionBusinessService, SubscriptionDetailsService}
+import uk.gov.hmrc.eoricommoncomponent.frontend.views.html.contact_address
+import uk.gov.hmrc.http.HeaderCarrier
+
+import javax.inject.{Inject, Singleton}
+import scala.concurrent.{ExecutionContext, Future}
+
+@Singleton
+class ContactAddressController @Inject() (
+  authAction: AuthAction,
+  subscriptionDetailsService: SubscriptionDetailsService,
+  subscriptionBusinessService: SubscriptionBusinessService,
+  cdsFrontendDataCache: SessionCache,
+  subscriptionFlowManager: SubscriptionFlowManager,
+  mcc: MessagesControllerComponents,
+  contactAddressView: contact_address
+)(implicit ec: ExecutionContext)
+    extends CdsController(mcc) {
+
+  def createForm(service: Service): Action[AnyContent] =
+    authAction.ggAuthorisedUserWithEnrolmentsAction { implicit request => _: LoggedInUserWithEnrolments =>
+      populateFormGYE(service)(false)
+    }
+
+  def reviewForm(service: Service): Action[AnyContent] =
+    authAction.ggAuthorisedUserWithEnrolmentsAction { implicit request => _: LoggedInUserWithEnrolments =>
+      populateFormGYE(service)(true)
+    }
+
+  private def populateFormGYE(
+    service: Service
+  )(isInReviewMode: Boolean)(implicit request: Request[AnyContent]): Future[Result] =
+    populateOkView(isInReviewMode = isInReviewMode, service)
+
+  def submit(isInReviewMode: Boolean, service: Service): Action[AnyContent] =
+    authAction.ggAuthorisedUserWithEnrolmentsAction { implicit request => _: LoggedInUserWithEnrolments =>
+      contactAddressDetailsYesNoAnswerForm
+        .bindFromRequest()
+        .fold(
+          formWithErrors =>
+            fetchContactDetails().map { contactAddressDetails =>
+              BadRequest(contactAddressView(contactAddressDetails, isInReviewMode, formWithErrors, service))
+            },
+          yesNoAnswer => saveAddress().flatMap(_ => locationByAnswer(isInReviewMode, yesNoAnswer, service))
+        )
+    }
+
+  private def fetchContactDetails()(implicit request: Request[AnyContent]): Future[AddressViewModel] =
+    cdsFrontendDataCache.subscriptionDetails flatMap { sd =>
+      sd.contactDetails match {
+        case Some(contactDetails) if contactDetails.street.isDefined =>
```

Comment:
Why are we just checking the street here?

Reply:
As the Contact Address information is amended to cache only after the address entry page, there is a check needed if any mandatory Address data exists already (here street is checked)  and accordingly retain values from ContactDetails cache to copy address data over to the Review and Submit method of contact-details page